teps/draft-20200715-access-control-via-sa.md

```diff
@@ -0,0 +1,249 @@
+
+However, periodic revisting of that stance quite possibly will be the mode of operation for Tekton contributors as the
+project moves forward. 
+
+### User Experience (optional)
+
+Highlighting and making it easy to learn the status of whether enforcement is enabled could be desirable from any of our 
+user experience platforms.
+
+Minimally, a new document on authorization within Tekton (analogous to the current authentication document) is needed
+to detail what is available, on or off by default, etc.
+
+### Performance (optional)
+
+An additional API server interaction now occurs with any PipelineRun or TaskRun for each referenced
+ClusterTask, Task, Pipeline.
+
+In the grand scheme of everything involved in executing a Pipeline, etc., not onerous.  Use of SARs in the 
+Kubernetes ecosystem would also corroborate this.
+
+That said, if the community ever gets contributions around performance analysis, comparisons around running 
+a set of Pipelines both with and without this feature enabled could be a candidate of interest for such an 
+endeavor.
+
+At the time of this writing, the author could not find anything in the community repositories around 
+performance benchmarks.  Any pointers from reviewers of course is appreciated.
+
+## Design Details
+
+For starters, a work in progress prototype of this proposal exists with this [PR](https://github.com/tektoncd/pipeline/pull/2797).
+
+At the time of this writing, the PR does is limited with respect to this proposal in that
+
+- it currently only checks cluster scoped resources
+- it currently does not yet have per namespace enablement
+
+Now, on the use of the ServiceAccount, either explicitly supplied or default if non supplied, for the permission check,
+a combination of advice from members of the Kubernetes sig-auth who I have talked to directly, as well as guidance
+from Kubernets sig-auth at the [2019 North American KubeCon Auth Deep Dive](https://www.youtube.com/watch?v=SFtHRmPuhEw),
+using the ServiceAccount for Pod security is better.  And at the end of the day, Tekton is ultimately about 
+translating high level objects and abstractions into Pods/Containers.
+
+The other choice is the "requesting user", which in the case of Tekton, could either be viewed as the 
+
+- the various Tekton controller/reconcilers ... this is who it is with Tekton today
+- or if you stored the user that created the TaskRun or PipelineRun object into that object via a mutating webhook
+and used that identity in subsequent validating webhooks (I explored and dismissed this approach on advice from 
+sig-auth members)
+
+But however the "requesting user" is set, if you listen to the regrets listed on the flawed auth model of the 
+PodSecurityPolicy admission plugin, from timestampls 19:11 to 21:14, you'll hear 
+
+- "Granting permissions to the requesting user is intuitive, but it breaks controllers"
+- "Dual mode weakens security ... you cannot have a privileged controller create pods on behalf of a user"
```

Comment:
Thanks so much for the pointer to the video here! I agree that using the user permissions is the intuitive choice, so I was kind of shocked to hear that it isn't recommended.

I just watched that section a few times, and I'm not convinced it actually applies in the case of Tekton. This might be worth a deeper discussion with those sig-auth members to clarify.

A couple big difference:
- one problem mentioned is that controllers break this model. The typical entity creating a pod is not a user, but another k8s controller (they used the deployment example). This means the creating user is almost always cluster-admin. In the case of Tekton, we don't create intermediate objects, we create pods directly.
- the other example is around container-breakouts using privileged containers and psp. I don't think that's really relevant here, or at least I can't see a connection.

In short, I'm (maybe naively) still hopeful that using the credentials of the requesting user, rather than the controller, would make this simpler and work for our case.

Reply:
Yeah, I'm not sure the intermediate steps needed or not needed to ultimately result in Pod creation matters.  Ultimately, we are creating Pods.

On Steps/Containers escalating in any fashion (given we exposed corev1.PodTemplate, corev1.Container) based on what either the requesting user (if we went down that path) or SA is allowed to do vs. what the Tekton controller(s) can do is pretty much exactly the point in my opinion.

All that said, I do agree that perhaps more direct conversations with folks in sig-auth, vs. our interpretations of their kube con discussions, would be ideal.  Common conversations with as many of the stakeholders and SMEs is always a better thing.  The person from my extended team who is on sig-auth is away from the office today, but I'll figure out when I can get some more cycles from him and report back.  Presumably we would want to attend an upcoming sig-auth meeting and present.  Certainly if you have thoughts / contacts / experience engaging with sig-auth @dlorenc let me know.

Certainly switching to include a requesting user is not a big change in the prototype I've posted (in fact I originally had a version that added requesting users to TaskRun and PipelineRun until my aforementioned team-mate on sig-auth talked me away from that).

Lastly, @skaegi and I have had some conversations on this (and he commented some on this topic in my precursor design document).  I believe my discussions with him (following my initial conversion to "use SAs, not requesting users") along with his personal experiences in his service has lead to a "use SAs, not requesting users" appraoch.

@skaegi - any chance you have cycles to chime in here?

thanks

Reply:
Note that regardless of what sig-auth says, I think it's important that the choices we make be grounded in specific use cases and understanding what scenarios we are envisioning as far as what happens without the functionality being proposed here.

Reply:
for sure ... I certainly was not trying to imply otherwise